pkg/controller/eventing/inmemory/clusterprovisioner/reconcile.go

```diff
@@ -0,0 +1,222 @@
+	logger   *zap.Logger
+}
+
+// Verify the struct implements reconcile.Reconciler
+var _ reconcile.Reconciler = &reconciler{}
+
+func (r *reconciler) InjectClient(c client.Client) error {
+	r.client = c
+	return nil
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	//TODO use this to store the logger and set a deadline
+	ctx := context.TODO()
+	logger := r.logger.With(zap.Any("request", request))
+
+	cp := &eventingv1alpha1.ClusterProvisioner{}
+	err := r.client.Get(ctx, request.NamespacedName, cp)
+
+	// The ClusterProvisioner may have been deleted since it was added to the workqueue. If so,
+	// there is nothing to be done.
+	if errors.IsNotFound(err) {
+		logger.Info("Could not find ClusterProvisioner", zap.Error(err))
+		return reconcile.Result{}, nil
+	}
+
+	// Any other error should be retried in another reconciliation.
+	if err != nil {
+		logger.Error("Unable to Get ClusterProvisioner", zap.Error(err))
+		return reconcile.Result{}, err
+	}
+
+	// Does this Controller control this ClusterProvisioner?
+	if !shouldReconcile(cp.Namespace, cp.Name) {
+		logger.Info("Not reconciling ClusterProvisioner, it is not controlled by this Controller", zap.String("APIVersion", cp.APIVersion), zap.String("Kind", cp.Kind), zap.String("Namespace", cp.Namespace), zap.String("name", cp.Name))
+		return reconcile.Result{}, nil
+	}
+	logger.Info("Reconciling ClusterProvisioner.")
+
+	err = r.reconcile(ctx, cp)
+	if err != nil {
+		logger.Info("Error reconciling ClusterProvisioner", zap.Error(err))
+		// Note that we do not return the error here, because we want to update the Status
+		// regardless of the error.
+	}
+
+	updateStatusErr := r.updateClusterProvisionerStatus(ctx, cp)
+	if updateStatusErr != nil {
+		logger.Info("Error updating ClusterProvisioner Status", zap.Error(updateStatusErr))
+		return reconcile.Result{}, updateStatusErr
+	}
+
+	return reconcile.Result{}, err
+}
+
+func IsControlled(ref *eventingv1alpha1.ProvisionerReference) bool {
+	if ref != nil && ref.Ref != nil {
+		return shouldReconcile(ref.Ref.Namespace, ref.Ref.Name)
+	}
+	return false
+}
+
+// shouldReconcile determines if this Controller should control (and therefore reconcile) a given
+// ClusterProvisioner. This Controller only handles in-memory buses.
+func shouldReconcile(namespace, name string) bool {
+	return namespace == "" && name == cpName
+}
+
+func (r *reconciler) reconcile(ctx context.Context, cp *eventingv1alpha1.ClusterProvisioner) error {
+	logger := r.logger.With(zap.Any("clusterProvisioner", cp))
+
+	// We are syncing one thing.
+	// 1. The K8s Service to talk to this in-memory bus.
+	//     - There is a single K8s Service for all requests going to this in-memory bus.
+
+	if cp.DeletionTimestamp != nil {
+		// K8s garbage collection will delete the dispatcher service, once this ClusterProvisioner
+		// is deleted, so we don't need to do anything.
+		return nil
+	}
+
+	err := r.createDispatcherService(ctx, cp)
```

Comment:
why dynamically create a service that is effectively a singleton when it could easily be part of the provisioner config yaml instead?

Reply:
A few revisions back, this controller would work on multiple in-memory ClusterProvisioners (e.g. in-memory-channel-foo, in-memory-channel-bar). In that world it made more sense to create the Service dynamically. 

The current revision dropped that ability, so this is vestigial. I'm trying to decide if I want to support multiple provisioners, thoughts? If we end up not wanting to support multiple provisioners this will likely get removed.

Reply:
I used (this commit)[https://github.com/adamharwayne/knative-eventing/commit/575995cf80fb9924d64023c16f8fc2b3ef98e124]. And when applying the altered [in-memory-channel.yaml](https://github.com/adamharwayne/knative-eventing/commit/575995cf80fb9924d64023c16f8fc2b3ef98e124#diff-5a216dd490d1162414a74cdd47d73616), I got:

```
error: error validating "STDIN": error validating data: 
ValidationError(Service.metadata.ownerReferences[0]): missing required field "uid" in 
io.k8s.apimachinery.pkg.apis.meta.v1.OwnerReference; if you choose to ignore these errors, turn 
validation off with --validate=false
2018/10/11 11:17:28 error executing "kubectl apply": exit status 1
```

If we want to make the K8s Service unowned, then we can switch to doing it via YAML. I prefer to have the owner reference and thus have the Controller make the K8s Service. Thoughts?

Reply:
Ping @scothis and @n3wscott